temporal-sdk/src/main/java/io/temporal/internal/worker/AsyncPoller.java

```diff
@@ -0,0 +1,387 @@
+          if (permit == null || shouldTerminate()) {
+            return;
+          }
+
+          pollerSemaphore.acquire();
+          pollerSemaphoreAcquired = true;
+          if (shouldTerminate()) {
+            return;
+          }
+          workerMetricsScope.counter(MetricsType.POLLER_START_COUNTER).inc(1);
+
+          SlotPermit finalPermit = permit;
+          CompletableFuture<T> pollRequest = asyncTaskPoller.poll(permit);
+          // Mark that we have made a poll request
+          pollRequestMade = true;
+          pollerBalancer.startPoll(asyncTaskPoller.getLabel());
+
+          pollRequest
+              .handle(
+                  (task, e) -> {
+                    pollerBalancer.endPoll(asyncTaskPoller.getLabel());
+                    if (e instanceof CompletionException) {
+                      e = e.getCause();
+                    }
+                    pollerSemaphore.release();
+                    pollScaleReportHandle.report(task, e);
+                    if (e != null) {
+                      uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), e);
+                      pollBackoffThrottler.failure(
+                          (e instanceof StatusRuntimeException)
+                              ? ((StatusRuntimeException) e).getStatus().getCode()
+                              : Status.Code.UNKNOWN);
+                      slotSupplier.releaseSlot(SlotReleaseReason.neverUsed(), finalPermit);
+                      return null;
+                    }
+                    if (task != null) {
+                      taskExecutor.process(task);
+                    } else {
+                      slotSupplier.releaseSlot(SlotReleaseReason.neverUsed(), finalPermit);
+                    }
+                    pollBackoffThrottler.success();
+                    return null;
+                  })
+              .exceptionally(
+                  throwable -> {
+                    log.error("Error while trying to poll task", throwable);
+                    return null;
+                  });
+        } catch (PollTaskAsyncAbort ab) {
+          abort = true;
+        } catch (Throwable e) {
+          if (e instanceof InterruptedException) {
+            // we restore the flag here, so it can be checked and processed (with exit) in finally.
+            Thread.currentThread().interrupt();
+          }
+          uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), e);
+        } finally {
+          // release the slot if it was acquired, but a poll request was not made
+          if (!pollRequestMade) {
+            if (permit != null) {
+              slotSupplier.releaseSlot(SlotReleaseReason.neverUsed(), permit);
+            }
+            if (pollerSemaphoreAcquired) {
+              pollerSemaphore.release();
+            }
+          }
+
+          if (shouldTerminate()) {
+            pollerBalancer.removePoller(asyncTaskPoller.getLabel());
+            abort = true;
```

Comment:
Seems like this abort = true and one above could just be `return`?

Reply:
You can't return inside a finally block